### ISSUE TITLE
Docs or example for writing gradfuns

### ISSUE THREAD

**Issue by thouis** [twitter/torch-autograd#67]
I have a workaround for specific application of #50 that involves using torch.eq(A, B), but returning the same type as A or B.  I was able to hack this into my system by putting this function into the torch module, and adding an entry in autograd.lua to define gradients (zero everywhere).

I'm sure there's a much cleaner way to do this, without making changes to autograd itself, but this was the easiest way I could discern.  A short example would be enough, I think.

**Comment by alexbw** [comment]
Is A a Tensor, and B a number? Post an example or make a PR, we can try to
iterate on your initial solution.

**Comment by thouis** [comment]
Both are tensors.  I was going to use them like this:

```
pooler = grad.nn.SpatialMaxPooling(2, 2)
unpooler = grad.nn.SpatialUpSamplingNearest(2)
pooled = pooler(input)
fmask = torch.feq(input, unpooler(pooled))
unpooled = torch.cmul(fmask, input)
```

This is the simple proof of concept:
```
torch.feq = function(a, b)
   return(a:eq(b):typeAs(a))
end

local grad = require 'autograd'

local W1 = torch.FloatTensor(1, 1, 10, 10):normal()
local W2 = W1 + 1  -- all different
W1[{1, 1, 1, 1}] = W2[{1, 1, 1, 1}]

local testeq = function(inputs)
   mask = torch.feq(inputs.W1, inputs.W2)
   product = torch.cmul(inputs.W1, mask)
   return torch.sum(product)
end

df = grad(testeq)
grads, val = df({W1=W1, W2=W2})
print(grads.W1)  -- should have 1 in the upper left corner
```

And in gradfuns.lua (in the wrapper for torch):
```
...
   module.gradient("feq", {
      function(g, ans, x, y)
         return util.zerosLike(g)
      end,
      function(g, ans, x, y)
         return util.zerosLike(g)
      end
   })
...
```

**Comment by alexbw** [comment]
Heyo, check out 39d7e1b. I've defined zero gradients for a lot of useful functions. Also, `torch.typeAs()` should work, too. So, you might try re-implementing your idea without the new `torch.feq` method, and seeing if it works as is.

**Comment by thouis** [comment]
Great!  Changed code to this:
```
local grad = require 'autograd'

local W1 = torch.FloatTensor(1, 1, 10, 10):normal()
local W2 = W1 + 1  -- all different
W1[{1, 1, 1, 1}] = W2[{1, 1, 1, 1}]

local testeq = function(inputs)
   local mask = torch.eq(inputs.W1, inputs.W2):typeAs(inputs.W1)
   local product = mask:cmul(inputs.W1)
   return torch.sum(product)
end

df = grad(testeq)
testeq({W1=W1, W2=W2})
grads, val = df({W1=W1, W2=W2})
print(grads.W1)  -- should have 1 in the upper left corner
```

Unfortunately, this code dies with:
```
luajit: test_unpool.lua:9: attempt to index local 'mask' (a nil value)
stack traceback:
        test_unpool.lua:9: in function 'fun'
        ...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:19: in function 'funOnly'
        ...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:114: in function 'df'
        test_unpool.lua:15: in main chunk
        [C]: in function 'dofile'
        ...ouis/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x0101d71bd0
```

It appears that typeAs is returning nil. If I take out the typeAs(), it dies at the cmul (trying to multiply bytes and floats).

**Comment by alexbw** [comment]
Calling "self" with functions is not supported in autograd. We don't have a
good error message for that yet. Use torch.cmul instead of mask:cmul and
torch.typeAs instead of res:typeAs.

**Comment by thouis** [comment]
Still no dice.  New code:
```
local grad = require 'autograd'

local W1 = torch.FloatTensor(1, 1, 10, 10):normal()
local W2 = W1 + 1  -- all different
W1[{1, 1, 1, 1}] = W2[{1, 1, 1, 1}]

local testeq = function(inputs)
   local mask = torch.typeAs(torch.eq(inputs.W1, inputs.W2), inputs.W1)
   local product = torch.cmul(inputs.W1, mask)
   return torch.sum(product)
end

df = grad(testeq)
testeq({W1=W1, W2=W2})
grads, val = df({W1=W1, W2=W2})
print(grads.W1)  -- should have 1 in the upper left corner
```

Error:
```
luajit: ...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:67: missing gradient for function torch.typeAs
stack traceback:
	[C]: in function 'error'
	...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:67: in function 'gradOnly'
	...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:124: in function 'df'
	test_unpool.lua:15: in main chunk
	[C]: in function 'dofile'
	...ouis/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x0104574bd0
```

**Comment by thouis** [comment]
A side note: torch.typeAs doesn't seem to be defined, though requiring autograd creates it?  By default it's a method on tensors, only.  The code above doesn't work if I remove the autograd-related code (the require, and the call to df).

**Comment by alexbw** [comment]
That's correct, there are some Torch methods which are only class methods, which is incompatible with autograd, so we define them ourselves.

**Comment by thouis** [comment]
Thanks for the clarification.

Does the code I posted above work for you?  I reinstalled torch & autograd, and it's still failing for me, but maybe there is something else broken in my environment.

**Comment by alexbw** [comment]
Btw the code you pasted above now works in the branch `awiltschko/grad-fixes`. It will be merged once I finish the gradients listed in #61. If you want to use it sooner, let me know, and I'll merge earlier. Closing.
